### DISCUSSION TITLE
Optimization of error signal

### DISCUSSION THREAD

**Discussion by NekSadam** [linien-org/linien#278] [Q&A]
Hello, 

I am currently trying to use _Linien_ in order to numerically obtain an error signal. After plugging in my AC spectroscopy signal, I am getting a error signal that I cannot optimize the way I would like to. I tried to modify the filters parameters on my own but without  success...

![CR stage 11 ](https://user-images.githubusercontent.com/104150875/172572622-8ccf5c78-4c72-4e6e-9867-99a6c1501bab.PNG)
 
So if anyone has the error signal calculation code or has any idea what is going on, feel free to give me a hand ! 

Thank you !

**Comment by NekSadam** [comment]
I should mention that this error signal comes from this spectroscopy signal : 

![CR Stage 4](https://user-images.githubusercontent.com/104150875/172573713-93d83cb9-49f3-46af-a855-6cfc2514cb19.PNG)

**Reply by bleykauf** [reply]
It is hard to understand what you are trying to do or what your setup looks like.

What is your AC spectroscopy signal? From the screenshot it looks like you are not modulating the signal (0 MHz modulation frequency). What is happening in the second screenshot? What do you mean by "numerically obtaining the error signal" and the "error signal calculation code"?

**Reply by NekSadam** [reply]
Hello, 

Thank your for your response and my apologies for having opened an issue topic, I was afraid it wouldn't be visible otherwise.

I am working using a Doppler-free rubidium 85/87 spectroscopy setup.

Here's a screenshot to have a better understanding of what is going on : 

![CR stage signal d'erreur 4](https://user-images.githubusercontent.com/104150875/172866360-01287375-29f0-4578-8b33-3d1a937af5d6.png)

So I was modulating the signal with a 1 kHz modulation frequency and an amplitude of 50 mVpp. The sweep signal's frequency is 7.5 Hz. 
What I meant by 'error signal calculation code' was rather how exaclty does _Linien_ compute the error signal with a given AC spectroscopy signal. 

Thanks again for your time !

**Reply by bleykauf** [reply]
Both your modulation frequency and amplitude are very low, though the latter is highly dependent on your setup. Typical modulation frequencies are on the order of the natural linewidth, depending on the modulation technique.

Modulation and demodulation take place on the FPGA. We use migen to generate the gateware, see the gateware folder.

**Reply by NekSadam** [reply]
In a previous analog setup my team and I used a modulation frequency of 7 kHz, way lower than the natural linewidth (6 MHz) and it worked just fine. 
Why wouldn't it work anyway ?

**Reply by bleykauf** [reply]
I still don't know what your setup looks like, are you using FMS or MTS?

Which parameters did you change in what range in order to optimize the error signal? What happens if you go to MHz modulation frequencies?

Regarding linewidth and modulation frequency:
For FMS: https://link.springer.com/article/10.1007/BF00688820
For MTS: http://www.braddye.com/technical/Frequency-Standards-Basics-and-Applications.pdf

**Reply by NekSadam** [reply]
Our all setup is made up of a simple rubidium 85/87 (observing the F = 3 -> F' rubidium 85 transition) Doppler-free spectroscopy setup. 
The injecting current in the laser diode is supplied by an LD controller with a bandwith of 200 kHz. We inject both the modulation and sweep signals coming out of the _Red Pitaya_ and plug them into an adder before sending in the sum into the LD controller. 
We send the monitor/error signals from the photodiode to FAST IN 1/2 and observe them with _Linien_. 

We have tried to modify every parameter possible in order to optimize the error signal. As the LD controller has a bandwith of 200 kHz, we cannot go any higher than that. We have noticed that modifying the sweep signal frequency appears to have an influence on the sampling frequency. 

I hope this has clarified what our problems were. 

Thanks again for your time !

**Reply by bleykauf** [reply]
Since you are sweeping and modulating with the same port, I don't see why you are using an additional adder. You can simply use the same channel for that. 

From the README: "You can also configure Linien for different setups, e.g. if you want to have the modulation frequency and the control on the same output." This can be done from the "General" tab.

**Reply by NekSadam** [reply]
We were not at first, but yes you're right we did send both the sweeping and modulating with the same port afterwards. 
Regardless, it shouldn't have any influence on the error signal now should it ?

**Reply by NekSadam** [reply]
EDIT : I placed a charging resistor after the photodiode (in parallel) and after adjusting the parameters I ended up with what looks like a decent error signal error. I used a 50 kHz modulation. 
Turns out the modulation might have been diminished by the photodiode. 

![image](https://user-images.githubusercontent.com/104150875/175542241-fa5909a7-2728-4b65-aefb-a7d3e6ee1a15.png)

Thank you for having taken the time to try to answer my questions ! 

I hope this discussion will be useful to some in the future :)

**Reply by bleykauf** [reply]
Happy that it works now!

Thanks for coming back and reporting your solution :)

**Reply by zhpshi** [reply]
Hello! It seems that I encountered the same problem. I modulated the strength of the B field around the rubidium cell. Then I observed the error signal same with the 3rd picture in this dialogue. Could you please tell me that the function of the charging resistor and the type of the PD you are using ? Thank you so much.

**Reply by NekSadam** [reply]
Hello zhpshi ! 

I don't know what is the B field you are refering to but I'll describe what did the trick for me. 

At first, my setup worked as follows : simple rubidium 85/87 cell -> photodiode (-> AC filtor) -> FAST IN 1/2 | FAST OUT 1 (modulation + sweep) -> laser diode (through a LD controller). Which gave me the 3rd picture you are talking about. 
Then, I put in between the photodiode and the FAST IN 1/2 a 10 kOhm in parallel configuration. It worked afterwards with a 50 kHz modulation frequency, 120 mVpp modulation amplitude and a 1.9 Hz frequency - filtering at 1 kHz. 

Feel free to keep asking me questions about my setup should you need further information. 

Good luck !

**Reply by zhpshi** [reply]
Hi Nek Sadam,

Thank you so much for your suggestions. We have observed the error signal after many times of trying.
